activerecord/lib/active_record/connection_adapters/postgresql/schema_statements.rb

```diff
@@ -354,17 +354,13 @@ def pk_and_sequence_for(table) #:nodoc:
 
         def primary_keys(table_name) # :nodoc:
           query_values(<<~SQL, "SCHEMA")
-            SELECT a.attname
-              FROM (
-                     SELECT indrelid, indkey, generate_subscripts(indkey, 1) idx
-                       FROM pg_index
-                      WHERE indrelid = #{quote(quote_table_name(table_name))}::regclass
-                        AND indisprimary
-                   ) i
-              JOIN pg_attribute a
-                ON a.attrelid = i.indrelid
-               AND a.attnum = i.indkey[i.idx]
-             ORDER BY i.idx
```

Comment:
Is this query deterministic order even if the ORDER BY is removed?

Reply:
Thanks for the comment. 

In my environment, it returns the primary key order as expected without order by but It does not look like deterministic since it does not `Sort` at the end of the query. 

Was thinking about adding `order by a.attnum` but it just shows the `column` order of the table, not the index key order.

```ruby
  create_table :countries_treaties, force: true, primary_key: [:country_id, :treaty_id] do |t|
    t.string :country_id, null: false
    t.string :treaty_id, null: false
  end
```

* Without `ORDER BY i.idx`


```sql
activerecord_unittest=#               SELECT a.attname
              FROM (
                     SELECT indrelid, indkey, generate_subscripts(indkey, 1) idx
                       FROM pg_index
                      WHERE indrelid = 'countries_treaties'::regclass
                        AND indisprimary
                   ) i
              JOIN pg_attribute a
                ON a.attrelid = i.indrelid
               AND a.attnum = i.indkey[i.idx];
  attname
------------
 country_id
 treaty_id
(2 rows)


activerecord_unittest=# explain SELECT a.attname
activerecord_unittest-#               FROM (
activerecord_unittest(#                      SELECT indrelid, indkey, generate_subscripts(indkey, 1) idx
activerecord_unittest(#                        FROM pg_index
activerecord_unittest(#                       WHERE indrelid = 'countries_treaties'::regclass
activerecord_unittest(#                         AND indisprimary
activerecord_unittest(#                    ) i
activerecord_unittest-#               JOIN pg_attribute a
activerecord_unittest-#                 ON a.attrelid = i.indrelid
activerecord_unittest-#                AND a.attnum = i.indkey[i.idx];
                                                         QUERY PLAN
-----------------------------------------------------------------------------------------------------------------------------
 Hash Join  (cost=329.65..357.92 rows=43 width=64)
   Hash Cond: ((pg_index.indrelid = a.attrelid) AND (pg_index.indkey[(generate_subscripts(pg_index.indkey, 1))] = a.attnum))
   ->  ProjectSet  (cost=0.27..13.30 rows=1000 width=34)
         ->  Index Scan using pg_index_indrelid_index on pg_index  (cost=0.27..8.29 rows=1 width=30)
               Index Cond: (indrelid = '207561'::oid)
               Filter: indisprimary
   ->  Hash  (cost=218.75..218.75 rows=7375 width=70)
         ->  Seq Scan on pg_attribute a  (cost=0.00..218.75 rows=7375 width=70)
(8 rows)

activerecord_unittest=#
```

* WITH `ORDER BY i.idx`
```sql
activerecord_unittest=# explain SELECT a.attname
              FROM (
                     SELECT indrelid, indkey, generate_subscripts(indkey, 1) idx
                       FROM pg_index
                      WHERE indrelid = 'countries_treaties'::regclass
                        AND indisprimary
                   ) i
              JOIN pg_attribute a
                ON a.attrelid = i.indrelid
               AND a.attnum = i.indkey[i.idx] ORDER BY i.idx;
                                                            QUERY PLAN
-----------------------------------------------------------------------------------------------------------------------------------
 Sort  (cost=359.09..359.20 rows=43 width=68)
   Sort Key: (generate_subscripts(pg_index.indkey, 1))
   ->  Hash Join  (cost=329.65..357.92 rows=43 width=68)
         Hash Cond: ((pg_index.indrelid = a.attrelid) AND (pg_index.indkey[(generate_subscripts(pg_index.indkey, 1))] = a.attnum))
         ->  ProjectSet  (cost=0.27..13.30 rows=1000 width=34)
               ->  Index Scan using pg_index_indrelid_index on pg_index  (cost=0.27..8.29 rows=1 width=30)
                     Index Cond: (indrelid = '207561'::oid)
                     Filter: indisprimary
         ->  Hash  (cost=218.75..218.75 rows=7375 width=70)
               ->  Seq Scan on pg_attribute a  (cost=0.00..218.75 rows=7375 width=70)
(10 rows)

activerecord_unittest=#
```

Reply:
Well, in the meantime I'm restoring this change to the current master one since it looks difficult (for me) to guarantee the order of index key, not table column order.